### DISCUSSION TITLE
How to scale one variable by another variable in multi-physics transient solver

### DISCUSSION THREAD

**Discussion by NorthMagic** [idaholab/moose#27210] [Q&A General]
Hello, I am solving on a coupling multi-physics transient problem. It is described by two coupled equations including two variables. At the end of each time step, one of the variables need to be scaled by another variable (solved in the current time step) for next time step. I didn't how to achieve it based on MOOSE framework, could you please give me some suggestions?

**Comment by GiudGiud** [comment]
Hello

nonlinear variables in MOOSE are supposed to only be computed by the nonlinear solve. As such, auxiliary kernels for example cannot set change their values. There are two workarounds:

- code a user object like "VariableScalingUO" that performs this scaling operation, executed on TIMESTEP_END, and does so by modifying the solution vector
- use a MultiApp setup in which you:

1. transfer the variable (on timestep_end) you want to scale to an auxiliary variable of exactly the same finite element family and order, in a multiapp with exactly the same mesh
2. use an auxkernel (parsedAux for example) to scale it there
3. transfer back from the auxvariable onto the nonlinear variable in the main app
Guillaume

**Reply by NorthMagic** [reply]
I am trying on the first way just as you mentioned, but I don't clearly know how to do it. I saw "Writing directly to coupled variables" about "Coupleable" in the MOOSE website. Should I get variables by "writableVariiable" and modify them by "setNodalValue"?

**Reply by GiudGiud** [reply]
For option 1 there are two options.
The coupleable interface writable values.
SetNodalValue will work only if you have a nodal variable (for example 1st order lagrange). Otherwise you will want to use setDofValue. For scaling, both should work

or accessing the solution vector directly from the nonlinear system

**Reply by NorthMagic** [reply]
Directly deal with solution vector from nonlinear system is a better choice. I know that `_fe_problem.getNonlinearSystem(0).solution()` can get the solution vector, dose it contain solutions of all coupled variables? How can I get the solution vector for each variable separately? I am so sorry for these minor questions.

**Reply by GiudGiud** [reply]
See the doxygen for all the apis available
```
https://mooseframework.inl.gov/docs/doxygen/moose/classNonlinearSystem.html
```
You ll want to use the variable number in the system
